<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Introduction and Principles of reproducibility.</title>
    <meta charset="utf-8" />
    <meta name="author" content="Emma Rand" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <script src="libs/fabric/fabric.min.js"></script>
    <link href="libs/xaringanExtra-scribble/scribble.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-scribble/scribble.js"></script>
    <script>document.addEventListener('DOMContentLoaded', function() { window.xeScribble = new Scribble({"pen_color":["#FF0000"],"pen_size":3,"eraser_size":30,"palette":[]}) })</script>
    <script src="libs/clipboard/clipboard.min.js"></script>
    <link href="libs/xaringanExtra-clipboard/xaringanExtra-clipboard.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-clipboard/xaringanExtra-clipboard.js"></script>
    <script>window.xaringanExtraClipboard(null, {"button":"Copy Code","success":"Copied!","error":"Press Ctrl+C to Copy"})</script>
    <link href="libs/xaringanExtra-extra-styles/xaringanExtra-extra-styles.css" rel="stylesheet" />
    <link rel="stylesheet" href="../css_files/emma.css" type="text/css" />
    <link rel="stylesheet" href="../css_files/emma-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Introduction and Principles of reproducibility.
]
.subtitle[
## White Rose BBSRC DTP Training: An Introduction to Reproducible Analyses in R.
]
.author[
### Emma Rand
]
.institute[
### University of York, UK
]

---













&lt;style&gt;
div.blue { background-color:#b0cdef; border-radius: 5px; padding: 20px;}
div.grey { background-color:#d3d3d3; border-radius: 0px; padding: 0px;}
&lt;/style&gt;


---

class: inverse

# Programme Overview

---

# What this training *is* and *is not*

Chosen topics are: foundational, widely applicable, and transferable conceptually.



.pull-left[

.font90[

**It is**
* An introduction to R for those without previous experience
* About using RStudio projects and good practice for code and project documentation and organisation
* An introduction to the tidyverse, RMarkdown, some more advanced data import
 

]

]

--

.pull-right[

.font90[

**It is not**
* An introduction to statistics  
* Magic  

]

]

---
# Programme overview

Modules, tutor-led or supported study. The selection of modules you undertake will depend on your previous experience.

.font60[

[1. Introduction and Principles of reproducibility](01_intro_and_principles_of_repro.html)  
Audience: Everyone
* Rationale for scripting  
* Why R?  
* Organisation of data  
* Organisation of analyses  

[2. A. Introduction to R and working with data](02_intro_to_r_and_working_with_data.html)  

Audience: Those without previous experience of R
* Finding your way round RStudio  
* Typing in data, doing some calculations on it, plotting it  
* Understanding the manual  
* Importing data: working directories and paths  
* Summarising and visualising with the [`tidyverse`](https://www.tidyverse.org/)  
* Installing and loading packages  

]

---
# Have experience?

.font60[

[2. B. Tidying data and the tidyverse including the pipe](04_tidying_data_and_the_tidyverse.html)  
Audience: For those with previous experience of R but little of 'tidy data' and the tidyverse
* Using the tidyverse including the pipe to link operations together.  
* Carrying out some common data tidying tasks such as reshaping, renaming and recoding variable and cleaning cell contents  



[2. C. Advanced data import](05_advanced_data_import.html)  
Audience: For those with previous experience of R and the tidyverse including the pipe."
* Understanding what matters in data import  
* Importing plain text and proprietary data formats stored locally and on the web  
* Carrying out some simple web scraping  
* Packages available for importing publicly accessible data from APIs  
]


---
# Programme overview

.font60[

[3. RStudio Projects](03_rstudio_projects.html)  
Audience: Those without experience of RStudio projects
* Organising your work in a logical, consistent and reproducible way using RStudio Projects  
* Writing code with relative paths given appropriate to your project organisation  
* Writing dataframes and figures to file  

[4. R Markdown for Reproducible Reports](06_r_markdown_for_reproducible_reports.html)  
Audience: For those with previous experience of R such as having done "Introduction to R and working with data."
* Making more advanced figures  
* Creating reproducible reports in a variety of output formats.  

]


---
class: inverse

# Data analysis


---
# Data analysis

How much of data analysis is using statistics?

Less than you probably think

~80% of your time on getting data, cleaning data, aggregating data, reshaping data, and exploring data using exploratory data analysis and data visualization.


---
# Reproducibility is key!

One definition
*"... obtaining consistent results using the same input data; computational steps, methods, and code; and conditions of analysis. This definition is synonymous with â€œcomputational reproducibility"*. (National Academies of Sciences, Engineering Medicine et al., 2019)

 Also see 

-  The Turing Way "Definitions for Reproducibility" 
-  National Science Foundation (Bollen Cacioppo et al., 2015)

---
# Reproducibility is key!

&lt;img src="pics/reproducible-matrix.jpg" width="700px" /&gt;

.font60[
How the Turing Way  defines reproducible research
]


---
# Who cares?

* Many high profile cases of work which did not reproduce e.g. Anil Potti unravelled by Baggerly and Coombes (2009) 

* Five selfish reasons to work reproducibly (Markowetz, 2015). Alternatively, see the [talk](https://youtu.be/yVT07Sukv9Q)


* **Will** become standard in Science and publishing e.g OECD Global Science Forum Building digital workforce capacity and skills for data-intensive science (OECD Global Science Forum, 2020)

---
# Open Science

&lt;img src="pics/Foster.png" width="700px" /&gt;
.font60[
By Petr Knoth and Nancy Pontika - https://en.wikipedia.org/wiki/Open_science#/media/File:Os_taxonomy.png, CC BY 3.0, https://commons.wikimedia.org/w/index.php?curid=61125075
]

FAIR - Findable, Accessible, Interoperable, Reusable (Wilkinson Dumontier et al., 2016)



---
class: inverse

# Rationale for scripting analysis


---
# Rationale for scripting analysis
Science is the generation of ideas, designing work to test them and reporting the results.

&lt;img src="pics/rationale1.png" width="600px" /&gt;

.pull-left[
Generating the results
]

.pull-right[
Analysing and reporting them
]
---

# Rationale for scripting analysis
We ensure reproducibility of laboratory and field work by planning and recording in lab books and using standard protocols. 


&lt;img src="pics/rationale2.png" width="600px" /&gt;

Even so replicating results can be hard.
---

# Rationale for scripting analysis
We ensure reproducibility of laboratory and field work by planning and recording in lab books and using standard protocols. 

&lt;img src="pics/rationale3.png" width="600px" /&gt;


Workflows for computational projects and the data analysis and reporting of other work can, and should, be 100% reproducible! Scripting is the way to achieve this.

---

# Rationale for scripting analysis

That reproducibility applies to all aspects of the data workflow.

&lt;img src="pics/rationale5.png" width="600px" /&gt;

--

From importing or collecting the data, processing it for analysis, building statistical models and communicating the methods and results.

--

These are usually iterative and that process of iteration (the development of the analysis) should also be captured.

---
class: inverse

# Why R?


---
# Why R?

Open source and free

--
 .......But so is Python
  
--

R has reputation for catering to users who do not see themselves as programmers, and  allowing them to slide gradually into programming.

![biologists](pics/biologist1.png)

--

Designed for data analysis and graphics - which means it is often easier to achieve those tasks in R than a general purpose programming language.

---
# Why R?

The R community is one of R's greatest assets, being vibrant, inclusive and supportive of users at all levels.

.pull-left[
* [#rstats](https://twitter.com/hashtag/rstats?lang=en) on twitter is very active
* [RForwards](https://forwards.github.io/about/) the widening participation task force &lt;sup&gt;1&lt;/sup&gt;
* [RLadies](https://rladies.org/) gender diversity promotion 
* [Hey! You there! You are welcome here](https://ropensci.org/blog/2017/06/23/community/)  

.font70[
.footnote[
1. I am member of the Core Team for Forwards
]
]

]

.pull-right[
&lt;img src="pics/welcome_to_rstats_twitter.png" width="400px" /&gt;

.font70[
.footnote[
Artwork by @allison_horst 
]
]

]

---
# Why R?

# Why R?

R Markdown is sometimes called R's "killer feature". Turns analyses into fully reproducible high quality reports, presentations and dashboards.

Can run Python and other languages in R Markdown documents

&lt;img src="pics/rmarkdown_wizards.png" width="500px" /&gt;

.font70[
.footnote[
Artwork by @allison_horst 
]
]
---
class: inverse

# Aspects of Reproducibility

---
# Organisation of data 

Within files should be 'tidy'

Tidy data adhere to a consistent structure which makes it easier to manipulate, model and visualize them. 

1. Each variable has its own column.  
2. Each observation has its own row.  
3. Each value has its own cell.  

Closely allied to the relational algebra of relational databases (Codd, 1990).  
Underlies the enforced rectangular formatting in SPSS, STATA and R's dataframe.  
The term 'tidy data' was popularised by Wickham (2014).   
There may be more than one potential tidy structure.  

---
# Tidy format



Suppose we had just 3 individuals in each of two populations:

.pull-left[

**Not tidy**

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:right;"&gt; A &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; B &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 12.4 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.6 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 11.2 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 11.6 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.1 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]


.pull-right[

**Tidy!**
&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; population &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; distance &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.4 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.2 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.6 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.6 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.1 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
]

---
# Organisation of analyses

* Project based approach, e.g., RStudio project or similar  
* directory structure 
* naming   
* commenting  
* readme  
* version control  

---
# Organisation of analyses

.pull-left[

**Directories**

* structured  
* systematic  
* repeatable  

**Naming**

* human and machine readable  
  * no spaces  
  * use snake/kebab case  
  * ordering: numbers (zero left padded), dates  
  * file extensions
]

.pull-right[
.code40[
```
-- stem_cell_rna_2019
   |__stem_cell_rna_2019.Rproj
   |__raw_ data
      |__2019-03-21_donor_1.csv
      |__2019-03-21_donor_2.csv
      |__2019-03-21_donor_3.csv
      |__2019-05-14_donor_1.csv
      |__2019-05-14_donor_2.csv
      |__2019-05-14_donor_3.csv
   |__processed_data
      |__all_long.txt
   |__figures
      |__01_volcano_donor_1_vs_donor_2.eps
      |__02_volcano_donor_1_vs_donor_3.eps
   |__functions
      |__01_file_import
      |__02_normalise.R
      |__theme_pca.R
      |__theme_volcano.R
   |__pics
      |__01_image.png
      |__01_image.png
   |__README.md
   |__refs
      |__r_refs.bib
      |__proj_refs.bib
   |__analyses
      |__01_data_processing.R
      |__02_exploratory.R
      |__03_modelling.R
      |__04_figures.R
      |__05_report.Rmd
```
]

]



---
# Organisation of analyses

## Further Reading

* "Ten simple rules for reproducible computational research" (Sandve Nekrutenko et al., 2013)  
* "Best practices for scientific computing" (Wilson Aruliah et al., 2014)  
* "Good enough practices in scientific computing" (Wilson Bryan et al., 2017)  
* "Excuse Me, Do You Have a Moment to Talk About Version Control?" (Bryan, 2018)  

---
# Summary

* The course is:
  * an introduction to reproducible analyses rather than statistics  
  * not enough, you need to practice!  
  * comprised of modules so you can opt out where you already have the skills  
* Scripting makes your work reproducible  
* Focus is on R but principles are widely applicable; use Python if you prefer  
* Recognising structure in your data and organising in 'tidy' format will pay dividends  
* The structured, systematic and consistent organisation of analyses will pay dividends



---
# References
.footnote[
.font60[
Slides made with with xaringan (Xie, 2019) and xaringanExtra (Aden-Buie, 2020)
]
]
.font60[
Aden-Buie, G. (2020). _xaringanExtra: Extras And Extensions for
Xaringan Slides_. R package version 0.2.3.9000. URL:
[https://github.com/gadenbuie/xaringanExtra](https://github.com/gadenbuie/xaringanExtra).

Baggerly, K. A. and K. R. Coombes (2009). "DERIVING CHEMOSENSITIVITY
FROM CELL LINES: FORENSIC BIOINFORMATICS AND REPRODUCIBLE RESEARCH IN
HIGH-THROUGHPUT BIOLOGY". In: _Ann. Appl. Stat._ 3.4, pp. 1309-1334.

Bollen, K., J. T. Cacioppo, et al. (2015). _Social, Behavioral, and
Economic Sciences Perspectives on Robust and Reliable Science_.
National Science Foundation.

Bryan, J. (2018). "Excuse Me, Do You Have a Moment to Talk About
Version Control?" In: _Am. Stat._ 72.1, pp. 20-27.

Codd, E. F. (1990). _The Relational Model for Database Management:
Version 2_. Boston, MA, USA: Addison-Wesley Longman Publishing Co.,
Inc.

Markowetz, F. (2015). "Five selfish reasons to work reproducibly". En.
In: _Genome Biol._ 16, p. 274.

National Academies of Sciences, Engineering, Medicine, et al. (2019).
_Understanding Reproducibility and Replicability_. National Academies
Press (US).
]

---
# References

.font60[
OECD Global Science Forum (2020). _Building digital workforce capacity
and skills for data-intensive science_. OECD.

Sandve, G. K., A. Nekrutenko, et al. (2013). "Ten simple rules for
reproducible computational research". En. In: _PLoS Comput. Biol._
9.10, p. e1003285.

Wickham, H. (2014). "Tidy Data". In: _Journal of Statistical Software,
Articles_ 59.10, pp. 1-23.

Wilkinson, M. D., M. Dumontier, et al. (2016). "The FAIR Guiding
Principles for scientific data management and stewardship". En. In:
_Sci Data_ 3, p. 160018.

Wilson, G., D. A. Aruliah, et al. (2014). "Best practices for
scientific computing". En. In: _PLoS Biol._ 12.1, p. e1001745.

Wilson, G., J. Bryan, et al. (2017). "Good enough practices in
scientific computing". En. In: _PLoS Comput. Biol._ 13.6, p. e1005510.

Xie, Y. (2019). _xaringan: Presentation Ninja_. R package version 0.12.
URL:
[https://CRAN.R-project.org/package=xaringan](https://CRAN.R-project.org/package=xaringan).
]
---
# Intro to Repro in R

Emma Rand  
[emma.rand@york.ac.uk](mailto:emma.rand@york.ac.uk)  
Twitter: [@er13_r](https://twitter.com/er13_r)   
GitHub: [3mmaRand](https://github.com/3mmaRand)  
blog: https://buzzrbeeline.blog/  

&lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;&lt;img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /&gt;&lt;/a&gt;&lt;br /&gt;&lt;span xmlns:dct="http://purl.org/dc/terms/" property="dct:title"&gt;Licensed under a &lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License&lt;/a&gt;.

Rand, E. (2021). White Rose BBSRC DTP Training: An Introduction to Reproducible Analyses in R (Version v1.1). https://doi.org/10.5281/zenodo.4701167
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
