---
title: "An Introduction to Reproducible Analyses in R: Tidying data and the tidyverse including the pipe."
subtitle: "White Rose BBSRC Doctoral Training Partnership (DTP) in Mechanistic Biology"
author: "Emma Rand"
institute: "University of York, UK"
date: "2020-19-03"
output:
  xaringan::moon_reader:
    css: [default, ../css_files/emma.css, ../css_files/emma-fonts.css]
    lib_dir: libs
    seal: false
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE,	
                      warning = FALSE)
options(htmltools.dir.version = FALSE)
```

```{r packages, include=FALSE}
library(RefManageR)
library(kableExtra)
library(tidyverse)
```


```{r, load-refs, include=FALSE, cache=FALSE}
BibOptions(check.entries = FALSE,
           bib.style = "authoryear",
           cite.style = "authoryear",
           style = "markdown",
           hyperlink = TRUE,
           dashed = FALSE,
           longnamesfirst = FALSE,
           max.names = 2)
myBib <- ReadBib("../refs.bib", check = FALSE)
```


class: inverse

# Introduction to Reproducibility in R 

## For those with previous experience: Tidying data and the tidyverse including the pipe.

### Emma Rand 

### University of York: 2020-03-19

.footnote[
Made with xaringan `r Cite(myBib, "xaringan")`  

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">White Rose BBSRC Doctoral Training Partnership (DTP) in Mechanistic Biology Analytics 1: Introduction to reproducible analyses in R</span> by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Emma Rand</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.
]


---
# Introduction

## Aims

The aim of this section is to introduce you to the tidyverse `r Cite(myBib, "Wickham2019-ml")` including the pipe, `%>%`  and some commonly applied data tidying operations. We are covering data tidying before the [Advanced Data Import]() section because some of the import methods will make use of tidyverse methods and generate tidy data structures.

## Learning outcomes 

The successful student will be able to:

* use pipes to link operations together.  
* carry out some common data tidying tasks - specify!
 

### `r emo::ji("clapper")` An instruction to do something!!

---
# Data tidying tasks

Tidying data includes reshaping it in to 'tidy' format but also other tasks such as:

* renaming variables for consistency  
* recoding variables  
* cleaning content for consistency with respect to valid values, missing values and NA  

--

Important:

* Keep the raw data exactly as it came to you and do not alter/edit.
* Script and document all tidying tasks.

---
# Tidyverse

The [tidyverse](https://www.tidyverse.org/) `r Cite(myBib, "R-tidyverse")` is "an opinionated" collection of R packages designed for data science. All packages share an underlying design philosophy, grammar, and data structures.  

Tidyverse is already installed on these PCs.

`r emo::ji("clapper")` Load the core tidyverse packages with:
```{r eval=FALSE}
library(tidyverse)
```

--
[Core packages](https://www.tidyverse.org/packages/) are: `ggplot2`, `dplyr`, `tidyr`, `readr`, `purr`, `tibble`, `stringr` and `forcats`

--

The tidyverse also includes many other packages with more specialised usage. They are not loaded automatically with `library(tidyverse)`and need their own `library()` calls. Examples are `readxl`, `haven`, `rvest` and `lubridate.`

---
# The pipe %>%

The pipe `%>%` operator from the `magrittr` package `r Cite(myBib, "magrittr")` is loaded with core tidyverse. 

It can improve code readability by:

* structuring sequences of data operations left-to-right (as opposed to from the inside and out),
* minimizing the need for intermediates, 
* making it easy to add steps anywhere in the sequence of operations.

--
Instead of using ` function(object)` you can use `object %>% function()`

--

This is useful when you have mulitple functions to apply.

---
# The pipe %>%

For example, suppose we want to apply a log-squareroot transformation<sup>1</sup> to some proprotion data.

.footnote[1. a transformation commonly applied to proportion data to make it less platykurtic $x^{t} = log(\sqrt{x})$
]

--

`r emo::ji("clapper")` Generate a random sample of ten proportions to work with:

```{r}
nums <- sample((1:100)/100, size = 10, replace = FALSE)
```

Two ways to apply this transformation are to:

1. nest the functions  
2. create intermediate variables  

---
# The pipe %>%

`r emo::ji("clapper")` Nest the `sqrt()` and `log()` functions:

```{r}
tnums <- log(sqrt(nums))
```


Disadvantage: Code must read from inside to outside. Increasingly difficult to read as number of functions increases. Also makes simple debugging harder.

--

`r emo::ji("clapper")` Create intermediate variables:

```{r}
sqrtnums <- sqrt(nums)
tnums <- log(sqrtnums)
```

Disadvantage: Extra variables you don't need. Not ideal if objects are large.

---
# The pipe %>%

Using the pipe avoids these by taking the output of one operation as the input of the next. 

The pipe has long been used by Unix operating systems (where the pipe operator is `|`). 
The R pipe operator is `%>%`, a short cut for which is ctrl-shift-M. 

--

`r emo::ji("clapper")` Use pipes to code the functions in sequence:
```{r}
tnums <- nums %>% 
  sqrt() %>% 
  log()
```

---
# The pipe %>%

More explicitly, this is:

```{r}
tnums <- nums %>% 
  sqrt(.) %>% 
  log(.)
```

Where `.` stands for the object being passed in. In most cases, you don't need to include it but some functions require you to (for example when arguments are optional or there is ambiguity over which argument is meant).

---
# Converting "wide" to "long"

Data commonly need to be reshaped from a format with more than one observation per row, for example, when there is a treatment group in each column. 

--

The data given in [biomass.txt](raw_data/biomass.txt) are taken from an experiment in which the insect pest biomass (g) was measured on plots sprayed  with water (control) or one of five different insecticides. Also in the data file are variables indicating the replicate number and the identity of the tray in which the plant was grown.

--

`r emo::ji("clapper")` Save a copy of this file to your `raw_data` folder and read it in.

```{r echo=FALSE}
file <- "raw_data/biomass.txt"
biomass <- read.table(file, header = T)

```

---
# Converting "wide" to "long"

`r emo::ji("clapper")` View the dataframe.

.scroll-output-height[

```{r echo=FALSE}

biomass

```

]


---
background-image: url(../pics/tidyr_spread_gather.png)
background-position: 50% 60%
background-size: 300px

# Converting "wide" to "long"

These data are in "wide" format and can be converting to "long" format using the `tidyr` package function `pivot_longer()`. 

--

By default `pivot_longer()` collects *all* the values into a single column (values_to) and creates a column to indicate the group (names_to). 

--

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
We want to gather the first 6 columns but the `rep_tray` column contents should be repeated.


---
# Converting "wide" to "long"

`r emo::ji("clapper")` Gather all the columns of biomass except `rep_tray`:

```{r}
biomass2 <- biomass %>% 
  pivot_longer(names_to = "spray", 
         values_to = "mass",
         -rep_tray)
```

The values will be in a column called `mass`, the treatments in a column called `spray`.


---
# Converting "wide" to "long"

`r emo::ji("clapper")` Write this dataframe to file to your `processed_data` folder.


```{r echo=FALSE}
file <- "processed_data/biomass2.txt"
write.table(biomass2, file = file, row.names = FALSE, quote = FALSE)

```

You can see the resulting file here:  [biomass2.txt](processed_data/biomass2.txt) 

.scroll-output-height[

```{r echo=FALSE}

as.data.frame(biomass2)

```

]

---
# Splitting column contents

We sometimes have single columns which contain more than one type of encoded information. 

--

UK Car number plates are an example - some parts of the number plate encode a year which you might want in an additional column for analyses by year. 

Dates or values with encoding for dates, experiment or treatment are common examples.

--

For the `biomass2` data we might wish to separate the replicate number from the tray identity and put them in two separate columns. 

We can do this with a 'regular expression' or [regex](https://en.wikipedia.org/wiki/Regular_expression). A regex defines a pattern for matching text. 

It's a big topic and there are many tutorials. I remember a few bits and google "how to match ... regex". 

[A quick reference](https://www.rexegg.com/regex-quickstart.html)

---
# Splitting column contents

The `extract()` function from the `tidyr` helps us achieve this.

We give:
* the names of the new coulmns we want to create
* the patterns matching the part of the `rep_tray` value we want to go in each column.

--

`r emo::ji("clapper")` Extract parts of `rep_tray` and put in new columns `replicate_number`, `tray_id`:

```{r}
biomass3 <- biomass2 %>% 
  extract(rep_tray, 
          c("replicate_number", "tray_id"),
          "([0-9]{1,2})\\_([a-z])")
```

--

* The patterns to save into columns are inside `( )`.  
* The pattern going into `replicate_number`, `[0-9]{1,2}`, means 1 or 2 numbers.  
* The pattern going into `tray_id`, `[a-z]` means one lowercase letter.
* the bit between the two `( )`, `\_` is a pattern that matches what is in `rep_tray` but is not to be saved. 

 
---
# Splitting column contents
.scroll-output-height[

```{r echo=FALSE}

as.data.frame(biomass3)

```

]

---
# Case study

## Overivew

You will now work through an example of some real data from [The Genever Group](https://www.geneverlab.info/). The arrangement and format of these data are typical of many protein and gene expression datasets so the processing is representative of that needed in a variety of situations.

--

The data are mass spectrometry data of the soluble protein fraction from five immortalised mesenchymal stromal cell (MSC) lines. The data are normalised protein abundances. Each row is a protein.

--

`r emo::ji("clapper")` Save a copy of [Y101_Y102_Y201_Y202_Y101-5.csv](raw_data/Y101_Y102_Y201_Y202_Y101-5.csv) file to your `raw_data` folder.

`r emo::ji("clapper")` You may wish to view it while reading the information on the next page.

---
# Case study

## Data description

The cells lines are Y101, Y102, Y201, Y202 and Y101.5 and there are three replicates for each cell line arranged in columns. Also in the file are columns for:

* the protein accession
* the number of peptides used to identify the protein
* the number of unique peptides used to identify the protein
* a measure of confidence in that identification
* the maximum fold change between the mean abundances of two cell lines (i.e., highest mean / lowest mean)
* a p value for a comparison test between the highest mean and lowest mean
* a q value (corrected p value) 
* a measure of the power of that test
* the cell line with the highest mean
* the cell line with the lowest mean
* the protein mass
* whether at least two peptides were detected for a protein.

---
# Case study

## Data Import

Column names are spreadover three rows but are primarily in the third row.
We can read in from the third row by skipping the first two. We can also use the `clean()` function from the `janitor` package to improve the column names.


`r emo::ji("clapper")` Read in the file using `read_csv()` from the tidyverse's `readr` package.
```{r}
# define file name
filesol <- "raw_data/Y101_Y102_Y201_Y202_Y101-5.csv"

# skip first two lines
sol <- read_csv(filesol, skip = 2) %>% 
  janitor::clean_names()

```

---
# Case study

## Data Import

`r emo::ji("eyes")` I used `read_csv()` because strings are treated as character variables, not factors. Whilst analysis and visualisation often require factor variables, any processing of strings is made much easier if they are characters.

`r emo::ji("eyes")` the `::` notation gives you access to a packages functions without first using the `library()` command. Useful when you want to use a single function from a package, or you need to specify which package when a function name is used in two loaded packages.

---
# Case study

## Filtering rows

This dataset includes bovine serum proteins from the medium on which the cells were grown which need to be filtered out.

--

We also filter out proteins for which fewer than 2 peptides were detected since we can not be confident about their identity. This is common practice for such proteomic data. 

--

`r emo::ji("clapper")` Keep rows of human proteins identified by more than one peptide:
```{r}
sol <- sol %>% 
  filter(str_detect(description, "OS=Homo sapiens")) %>% 
  filter(x1pep == "x")
```
`r emo::ji("eyes")` `str_detect(string, pattern)` returns a logical vector according to whether 'pattern' is found in 'string'.

---
# Case study

## Processing cells contents

It would be useful to extract the genename from the description and put it in a column.

One entry from the description column looks like this:

.scroll-output-width[

```{r}
sol$description[1]
```
]

--

The genename is after `GN=`. We need to extract the part of the string with the genename and put it in a new column. 

A way to problem-solve your way through this is work with *one* value carrying out one operation at a time until you've worked out what to do before implementing on an entire column.

---
# Case study

## Processing cells contents

### One step at a time on one value

`r emo::ji("clapper")` Extract the first value of the description to work with:
```{r}
one_description <- sol$description[1]
```

--

`r emo::ji("clapper")` Extract the part of the string after `GN=` using a regex:
```{r}

one_genename <- one_description %>%
  str_extract("GN=[^\\s]+")
one_genename
```

Explanation on next slide.

---
# Case study

## Processing cells contents

### One step at a time on one value


* `[ ]` means some characters
* `^` means 'not' when inside `[ ]`
* `\s` means white space  
* the `\` before is an escape character to indicate that the next character, `\` should not be taken literally (because it's part of `\s`)  
* `+` means one or more  

--

So `GN=[^\\s]+` means `GN=` followed by one or more characters that are not whitespace. This means the pattern stops matching at the first white space after "GN=".

---
# Case study

## Processing cells contents

### One step at a time on one value

We're close. Now we will drop the `GN` part by spliting the string on the `=`:

```{r}
# extract the part of the string
one_genename <- one_description %>%
  str_extract("GN=[^\\s]+") %>% 
  strsplit(., "=", fixed = TRUE)
one_genename
```

Note we need to use `.` to represent the object being passed in, "GN=AHNAK" in this case. This is because the function definition of [`strsplit()`(https://www.rdocumentation.org/packages/base/versions/3.6.1/topics/strsplit]. The `split` argument would be mistaken for the `x` argument.

The output value of `strsplit()` is a two element list of the strings before and after the "=". We want the second element of the list. An often easier way to access lists where each item in the list is a single string is to turn it into a character vector. The `unlist()` function achieves this:

```{r}
# extract the part of the string
one_genename <- one_description %>%
  str_extract("GN=[^\\s]+") %>% 
  strsplit(., "=", fixed = TRUE) %>% 
  unlist() 
one_genename
```
The to access the second element of the the character vector we need to use the `.` again:
```{r}
# extract the part of the string
one_genename <- one_description %>%
  str_extract("GN=[^\\s]+") %>% 
  strsplit(., "=", fixed = TRUE) %>% 
  unlist()  %>% 
  .[2]
one_genename
```

We know how to get the genename for a single value and now need to apply this to every value in the column. This can be achieved with a `for` loop:

```{r}
# add column for genename
for (i in 1:length(sol$description)) { 
  sol$genename[i] <- sol$description[i]  %>%
    str_extract("GN=[^\\s]+")  %>% 
    strsplit(., "=", fixed = TRUE)  %>%
    unlist()   %>% 
    .[2]}
```


# Exercises
1. Can you add a column for the top protein identifier. This is the first Uniprot ID after the "1::" in the `accession` column. I suggest splitting on the semi-colon, then the double colon. Develop the solution by using a suitable entry from the accession column (i.e., not the first one as it only has one entry).

```{r echo=FALSE}
# add column for top protein identifier.
# 
for (i in 1:length(sol$accession)) {
  sol$uniprotid[i] <- sol$accession[i] %>%
    strsplit(., ";", fixed = TRUE)  %>%
    unlist()   %>% 
    .[1] %>%
    strsplit(., "::", fixed = TRUE)  %>%
    unlist()   %>%
    .[2]}
```

2. Can you create a second dataframe, `sol2` in which the protein abundances are in a single column and the cell lineage is indicated in another. All the other variables should also be in the new data frame.
```{r echo=FALSE}
# # you may have list the columns you did not wanted to gather
# # but here's another trick
# to_gather <- colnames(sol)[(colnames(sol) %like% "y")]
# to_gather
# sol2 <- sol %>% gather(key = lineage,
#                        value = abundance,
#                        to_gather)


```
3. Can you create separate columns in `sol2` for the cell line and the replicate?

```{r echo=FALSE}
# sol2 <- sol2 %>% 
#   extract(lineage, 
#           c("line", "rep"),
#           "(y[0-9]{3,4})\\_([a-c])")
```




---
# References

```{r refs, echo=FALSE, results="asis"}
PrintBibliography(myBib)
```



---
class: inverse

# Congratulations! Keep practising!

Emma Rand 

emma.rand@york.ac.uk

Twitter: [@er13_r](https://twitter.com/er13_r) 

github: [3mmaRand](https://github.com/3mmaRand)

blog: https://buzzrbeeline.blog/
<br>

.footnote[
<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">White Rose BBSRC Doctoral Training Partnership (DTP) in Mechanistic Biology Analytics 1: Introduction to reproducible analyses in R</span> by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Emma Rand</span> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.
]






